text/0000-stable-simd.md

```diff
@@ -0,0 +1,720 @@
+**thousands** of functions to the standard library in a stable manner. It's
+infeasible for any one person (or even the entire libs team) to scrutinize all
+functions and assess whether the correct signature is applied (aka was it
+`i8x16` or `i16x8`?)
+
+Furthermore not all intrinsics from Intel actually have an interpretation with
+one of the portable types. For example some intrinsics take an integer constant
+which when 0 interprets the input as `u8x16` and when 1 interprets it as
+`u16x8` (as an example). This effectively means that there *isn't* a correct
+choice in all situations for what portable type should be used.
+
+Consequently it's proposed that instead of portable types the exact vendor types
+are used in all vendor intrinsics. This provides us a much easier route to
+stabilization ("make sure it's what Intel says") along with no need to interpret
+what Intel does and attempt to find the most appropriate type.
+
+There is interest by both current `stdsimd` maintainers and users to expose
+a "better-typed" SIMD API in crates.io that builds on top of the intrinsics
+proposed for stabilization here.
+
+## Stabilizing SIMD implementation details
+
+Another alternative to the bulk of this RFC is allowing more raw access to the
+internals of LLVM. For example stabilizing `#[repr(simd)]` or the ability to
+write `extern "platform-intrinsics" { ... }` or `#[link_llvm_intrinsic...]`.
+This is certainly a *much* smaller surface area to stabilize (aka not
+thousands of intrinsics).
+
+This avenue was decided against, however, for a few reasons:
+
+* Such raw interfaces may change over time as they simply represent LLVM as a
+  current point in time rather than what LLVM wants to do in the future.
+* Alternate implementations of rustc or alternate rustc backends like Cretonne
+  may not expose the same sort of functionality that LLVM provides, or
+  implementing the interfaces may be much more difficult in alternate backends
+  than in LLVM's.
+
+As a result, it's intended that instead of exposing raw building blocks (and
+allowing `stdsimd` to live on crates.io) we'll instead pull in `stdsimd` to the
+standard library and expose it as the stable interface to SIMD in Rust.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+There's a number of unresolved questions around stabilizing SIMD today which
+don't pose serious blockers and may also wish to be considered open bugs rather
+than blocking stabilization:
+
+## Relying on unexported LLVM APIs
+
+The static detection performed by `cfg!` and `#[cfg]` currently relies on a
+[Rust-specific patch to LLVM][llvm-patch]. LLVM internal knows all about
+hierarchies of features and such. For example if you compile with `-C
+target-feature=+avx2` then `cfg!(target_feature = "sse2")` also needs to resolve
+to `true`. Rustc, however, does not know about these features and relies on
+learning this information through LLVM.
```

Comment:
If it's *only* the inclusion relations between target features, that's not a lot of information to duplicate (and not very arcane either). Seems fine to me to just include this information in rustc.

Reply:
I do believe that's mostly all we need, yeah! The only other thing I can think of is `-C target-cpu=native` but if we have feature detection in libstd then we may not need to replicate how LLVM does that detection, we could do it ourselves!

I'd just not trust myself to translate this logic, but I'm also personally ok leaving this as an open bug as it should be easy enough (I think) to fix down the line!